// @pnpm/lockfile-file/lib/sortLockfileKeys.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sortLockfileKeys = void 0;
const sort_keys_1 = __importDefault(require("sort-keys"));
const ORDERED_KEYS = {
    resolution: 1,
    id: 2,
    name: 3,
    version: 4,
    engines: 5,
    cpu: 6,
    os: 7,
    deprecated: 8,
    hasBin: 9,
    prepare: 10,
    requiresBuild: 11,
    bundleDependencies: 12,
    peerDependencies: 13,
    peerDependenciesMeta: 14,
    dependencies: 15,
    optionalDependencies: 16,
    transitivePeerDependencies: 17,
    dev: 18,
    optional: 19,
};
const ROOT_KEYS_ORDER = {
    lockfileVersion: 1,
    neverBuiltDependencies: 2,
    overrides: 3,
    packageExtensionsChecksum: 4,
    specifiers: 10,
    dependencies: 11,
    optionalDependencies: 12,
    devDependencies: 13,
    importers: 14,
    packages: 15,
};
function compareWithPriority(priority, left, right) {
    const leftPriority = priority[left];
    const rightPriority = priority[right];
    if (leftPriority && rightPriority)
        return leftPriority - rightPriority;
    if (leftPriority)
        return -1;
    if (rightPriority)
        return 1;
    return left.localeCompare(right);
}
function sortLockfileKeys(lockfile) {
    const compareRootKeys = compareWithPriority.bind(null, ROOT_KEYS_ORDER);
    if (lockfile.importers != null) {
        lockfile.importers = sort_keys_1.default(lockfile.importers);
        for (const importerId of Object.keys(lockfile.importers)) {
            lockfile.importers[importerId] = sort_keys_1.default(lockfile.importers[importerId], {
                compare: compareRootKeys,
                deep: true,
            });
        }
    }
    if (lockfile.packages != null) {
        lockfile.packages = sort_keys_1.default(lockfile.packages);
        for (const pkgId of Object.keys(lockfile.packages)) {
            lockfile.packages[pkgId] = sort_keys_1.default(lockfile.packages[pkgId], {
                compare: compareWithPriority.bind(null, ORDERED_KEYS),
                deep: true,
            });
        }
    }
    for (const key of ['specifiers', 'dependencies', 'devDependencies', 'optionalDependencies']) {
        if (!lockfile[key])
            continue;
        lockfile[key] = sort_keys_1.default(lockfile[key]);
    }
    return sort_keys_1.default(lockfile, { compare: compareRootKeys });
}
exports.sortLockfileKeys = sortLockfileKeys;
